controllers/batchrelease/batchrelease_controller.go

```diff
@@ -0,0 +1,230 @@
+// +kubebuilder:rbac:groups=apps,resources=replicasets/status,verbs=get;update;patch
+
+// Reconcile reads that state of the cluster for a Rollout object and makes changes based on the state read
+// and what is in the Rollout.Spec
+func (r *BatchReleaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	release := new(v1alpha1.BatchRelease)
+	err := r.Get(context.TODO(), req.NamespacedName, release)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	klog.Infof("Begin to reconcile BatchRelease(%v/%v), release-phase: %v", release.Namespace, release.Name, release.Status.Phase)
+
+	// finalizer will block the deletion of batchRelease
+	// util all canary resources and settings are cleaned up.
+	reconcileDone, err := r.handleFinalizer(release)
+	if reconcileDone {
+		return reconcile.Result{}, err
+	}
+
+	// set the release info for executor before executing.
+	r.executor.SetReleaseInfo(release)
+
+	// executor start to execute the batch release plan.
+	startTimestamp := time.Now()
+	result, currentStatus := r.executor.Do()
+
+	defer func() {
+		klog.InfoS("Finished one round of reconciling release plan",
+			"BatchRelease", client.ObjectKeyFromObject(release),
+			"phase", currentStatus.Phase,
+			"current-batch", currentStatus.CanaryStatus.CurrentBatch,
+			"current-batch-state", currentStatus.CanaryStatus.CurrentBatchState,
+			"reconcile-result ", result, "time-cost", time.Since(startTimestamp))
+	}()
+
+	return result, r.updateStatus(ctx, release, currentStatus)
+}
+
+func (r *BatchReleaseReconciler) updateStatus(ctx context.Context, release *v1alpha1.BatchRelease, newStatus *v1alpha1.BatchReleaseStatus) error {
+	var err error
+	defer func() {
+		if err != nil {
+			klog.Errorf("Failed to update status for BatchRelease(%v), error: %v", client.ObjectKeyFromObject(release), err)
+		}
+	}()
+
+	// observe and record the latest changes for generation and release plan
+	newStatus.ObservedGeneration = release.Generation
+	newStatus.ObservedReleasePlanHash = hashReleasePlanBatches(&release.Spec.ReleasePlan)
+
+	// do not retry
+	if !reflect.DeepEqual(release.Status, *newStatus) {
+		releaseClone := release.DeepCopy()
+		releaseClone.Status = *newStatus
+		return r.Status().Update(ctx, releaseClone)
+	}
+
+	return err
+}
+
+func (r *BatchReleaseReconciler) handleFinalizer(release *v1alpha1.BatchRelease) (bool, error) {
+	var err error
+	defer func() {
+		if err != nil {
+			klog.Errorf("Failed to handle finalizer for BatchRelease(%v), error: %v", client.ObjectKeyFromObject(release), err)
+		}
+	}()
+
+	// remove the release finalizer if it needs
+	if !release.DeletionTimestamp.IsZero() &&
```

Comment:
These conditions have considered whether the underlying resources have been cleaned up. If the controller crash before cleaning up, it will retry to clean up when it is resumed.